React/forms.md

```diff
@@ -2,24 +2,280 @@
-- Handle the event of a form submission in a component
 - Dynamically provide user feedback as they complete a form
+- Handle the event of a form submission in a component
 
 ### Overview
 Since we've already learn how event handling works in React, we should be able to take the application of this knowledge one step further to allow us to submit a set of form data. Forms work a bit different than other HTML elements because they maintain some of their own state.
 
 Additionally, one of the most powerful aspects of using any JavaScript framework is to provide some dynamic user interaction. We'll see here how to integrate ongoing user feedback, so the user does not have to wait until the form is submitted to know that they might have issues with the data.
 
+## A Regular HTML Form
+
+A normal HTML form to add a student to a list might look like this:
+
+```html
+<form id="new-student" >
+  <label for="name">Student Name</label>
+  <input name="name" type="text">
+  <label for="email">Student Email</label>
+  <input name="email" type="text">
+  <input type="submit" value="Add Student">
+</form>
+```
+
+When a user enters text in the form controls the DOM stores the values of the form fields.  However we can do a lot more combining it with React.
+
+First we will build a React component to manage creating new students.
+
+```javascript
+import React, { Component } from 'react';
+import  './new_student_form.css';
+
+class NewStudentForm extends Component {
+  constructor(props) {
+    super(props);
+}
+render() {
+    return (
+      <div>
+        <form className="new-student-form">
+          <div>
+            <label htmlFor="name">Name:</label>
+            <input
+              name="name" />
+          </div>
+          <div>
+            <label htmlFor="email">Email:</label>
+            <input name="email" />
+          </div>
+          <input className="button success" type="submit" value="Add Student" />
+        </form>
+      </div>
+    );
+  }
+}
+
+export default NewStudentForm;
+```
+
+Now we have a React component to render the form.  It functions just like the HTML form, but can be imported into other React containers.  We term the name and email input fields _Uncontrolled components_ because the DOM is managing the values in the fields.  However now that we have them in a React component we can modify the form controls to have React monitor the input fields and handling the form submission.  To do so we will turn the inputs for name and email into _controlled components_.
+
+We can add this to the constructor:
+
+```javascript
+//  new_student_form.js
+...
+constructor(props) {
+  this.state = {
+    name: '',
+    email: '',
+  }
+}
+```
+
+The `NewStudentForm` component will track the `name` and `email` as part of it's state.
+
+We can then manage the input components by setting the value of the form component to match the `NewStudentForm`'s state and adding an event handler to respond to changes in the input value.
```

Comment:
I don't think we refer to the elements in JSX as components. I think the input and the form are elements or fields not components.

Reply:
I think the confusion might be from the fact that this `NewStudentForm` component is becoming a controlled component overall - but I don't think we refer to each individual field using the same vocabulary.